cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ProgramGenerationStage.java

```diff
@@ -120,7 +127,9 @@ public Location call() throws Exception {
             }
             Location output = programDir.append(String.format("%s.jar", spec.getName()));
             Id.Program programId = Id.Program.from(input.getId(), type, spec.getName());
-            return ProgramBundle.create(programId, bundler, output, spec.getClassName(), appSpec);
+            Location programLocation = ProgramBundle.create(programId, bundler, output, spec.getClassName(), appSpec);
+            authorizer.grant(programId.toEntityId(), SecurityRequestContext.toPrincipal(), ImmutableSet.of(Action.ALL));
+            return programLocation;
```

Comment:
Did it this way because we want to grant access only if ``ProgramBundle.create`` succeeds